tensorflow/core/kernels/histogram_op.cc

```diff
@@ -69,6 +72,79 @@ struct HistogramFixedWidthFunctor<CPUDevice, T, Tout> {
   }
 };
 
+#ifdef TENSORFLOW_USE_SYCL
+
+namespace {
+
+// Generate a matrix wich values are the index of the column.
+template <typename T>
+struct ColIndicesGenerator {
+  inline T operator()(const Eigen::array<Eigen::DenseIndex, 2>& idx) const {
+    return idx[1];
+  }
+};
+
+} //namespace
+
+template <typename T, typename Tout>
+struct HistogramFixedWidthFunctor<SYCLDevice, T, Tout> {
+  static Status Compute(OpKernelContext* context,
+                        const typename TTypes<T, 1>::ConstTensor& values,
+                        const typename TTypes<T, 1>::ConstTensor& value_range,
+                        int32 nbins, typename TTypes<Tout, 1>::Tensor& out) {
+    const SYCLDevice& d = context->eigen_device<SYCLDevice>();
+
+#ifdef TENSORFLOW_SYCL_NO_DOUBLE
+    using InternalT = T;
+#else
+    using InternalT = double;
+#endif  // TENSORFLOW_SYCL_NO_DOUBLE
+
+    auto values_size = values.size();
+    if (values_size == 0) {
+      out.device(d) = out.constant(0);
+      return Status::OK();
+    }
+
+#if !defined(EIGEN_HAS_INDEX_LIST)
+    Eigen::DSizes<Eigen::Index, 1> sum_dim(0);
+    Eigen::DSizes<Eigen::Index, 2> values_size_by_one({values_size, 1});
+    Eigen::DSizes<Eigen::Index, 2> one_by_nbins({1, nbins});
+#else
+    Eigen::IndexList<Eigen::type2index<0>> sum_dim;
+    Eigen::IndexList<Eigen::Index, Eigen::type2index<1>> values_size_by_one;
+    values_size_by_one.set(0, values_size);
+    Eigen::IndexList<Eigen::type2index<1>, Eigen::Index> one_by_nbins;
+    one_by_nbins.set(1, nbins);
+#endif
+
+    const InternalT step =
+      static_cast<InternalT>(value_range(1) - value_range(0)) /
+      static_cast<InternalT>(nbins);
+
+    // The calculation is done by finding the slot of each value in `values`.
+    // With [a, b]:
+    //   step = (b - a) / nbins
+    //   (x - a) / step
+    // , then the entries are mapped to output.
+    auto index_to_bin =
+        ((values.cwiseMax(value_range(0)) - values.constant(value_range(0)))
+            .template cast<InternalT>() / step)
+          .template cast<int32>()
+          .cwiseMin(nbins - 1);
+    auto index_to_bin_2d = index_to_bin.reshape(values_size_by_one)
```

Comment:
Shouldn't we use `.eval()` after reshaping?

Reply:
Why? eval will force to create a temporary buffer here.